### PR TITLE
update README.md, PR template and make sure all folders contain policies with the same name

### PR DESCRIPTION
**Pull request by guohaoyu110**
## Related Issue(s)
Close: https://github.com/kyverno/policies/issues/553

## Description
update readme to include artifact-pkg.yml template, artifact-pkg.yml metadata instructions and include updated PR template
and make sure all folders contain policies with the same name
- [x] All Kyverno policies must be in individual folders.
- [x] All folders contain policies with the same name.
-  [x] A template/example policy metadata file (artifacthub-pkg.yml) is present in the [README.md](https://github.com/kyverno/policies/blob/main/README.md) file.
- [x] The [README.md](https://github.com/kyverno/policies/blob/main/README.md) file is updated in the [Contribution section](https://github.com/kyverno/policies#contribution) to have instructions on how to add the metadata file and complete its contents.
- [x] The [PR template](https://github.com/kyverno/policies/blob/main/.github/PULL_REQUEST_TEMPLATE.md) is updated to include a checkbox to remind future contributors about the metadata file.

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.

### CODE REVIEW

**Comment by guohaoyu110** [comment]
@chipzoller  updated, please review. Thank you.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -19,3 +19,4 @@ them, don't hesitate to ask. We're here to help! This is simply a reminder of wh
 
 - [] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
 - [] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
+- [] I have added artifact-pkg.yml file with the policy metadata and refer to the [template](https://github.com/artifacthub/hub/blob/master/docs/metadata/artifacthub-pkg.yml)
```
Comment (chipzoller) [inline .github/PULL_REQUEST_TEMPLATE.md:22]: Name is wrong.

File: README.md [hunk 2]
```diff
@@ -58,6 +58,44 @@ spec:
             - Resource
 ```
 
+
```
Comment (chipzoller) [inline README.md:61]: Add this under a new heading called `### Artifact Hub`

Comment (chipzoller) [reply README.md:61]: Add a new bullet point where L30 currently is that states:

"* Create the `artifacthub-pkg.yml` file in the same directory as your policy. See the Artifact Hub section below for more details on its contents."

Comment (chipzoller) [reply README.md:61]: Find the appropriate bullet above in the same Contribution heading to state the following things:

1. A dedicated folder must be created for each policy.
2. The folder must be named the same as the policy.
3. Use dashes and not underscores.

File: README.md [hunk 3]
```diff
@@ -58,6 +58,44 @@ spec:
...
+When adding a new policy, please follow these guidelines:
+
+1. Create a new folder with the name of the policy.
+2. Add the Kyverno policy YAML file to the folder.
```
Comment (chipzoller) [inline README.md:65]: Eliminate and incorporate as per the above comment.

File: README.md [hunk 4]
```diff
@@ -58,6 +58,44 @@ spec:
...
+3. Add an `artifacthub-pkg.yml` metadata file to the folder. See the [Template Policy Metadata File](https://github.com/artifacthub/hub/blob/master/docs/metadata/artifacthub-pkg.yml) section for an example and instructions on how to complete the contents of this file.
```
Comment (chipzoller) [inline README.md:66]: Change second sentence to:

"See an example metadata file for Kyverno policies below and customize per the comments."

File: README.md [hunk 5]
```diff
@@ -58,6 +58,44 @@ spec:
...
+
+```yaml
+# artifacthub-pkg.yml
```
Comment (chipzoller) [inline README.md:69]: Remove

File: README.md [hunk 6]
```diff
@@ -58,6 +58,44 @@ spec:
...
+  
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+
+annotations:
+  kyverno/category: refer to policies.kyverno.io/category in policy file 
+  kyverno/kubernetesVersion: refer to kyverno.io/kubernetes-version in policy file
+  kyverno/subject: refer to policies.kyverno.io/subject in policy file
+  
+```
+
```
Comment (chipzoller) [inline README.md:97]: Eliminate excess new lines. Only a single new line should separate code block fencing from headings.

File: README.md [hunk 7]
```diff
@@ -58,6 +58,44 @@ spec:
...
+  Copy the content of you policy description here.
+  
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+
+annotations:
+  kyverno/category: refer to policies.kyverno.io/category in policy file 
+  kyverno/kubernetesVersion: refer to kyverno.io/kubernetes-version in policy file
+  kyverno/subject: refer to policies.kyverno.io/subject in policy file
+  
+```
```
Comment (chipzoller) [inline README.md:96]: Please replace all this with one of the examples you created from the Velero folder. In addition, add to the end of each line a comment which describes what goes in each value.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Comment by chipzoller** [comment]
From the check list above:

```
All Kyverno policies must be in individual folders.
```

This is _not_ addressed in this PR. As mentioned the other day, there are many policies in the `other` folder which live outside of folders. All policies in this repo must be in dedicated folders which use dashes over underscores and where the folder is the same as the policy name.

**Comment by chipzoller** [comment]
Please also go through ALL policies in this repo and ensure that the policies match the folder name. I have only pointed out two in the Kasten folder, but there may be others as I did not do this for you.

**Comment by chipzoller** [comment]
Please also update the description since you are covering other tasks than just those you mentioned.

**Comment by guohaoyu110** [comment]
@chipzoller  updated all, please review. Thank you.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -19,3 +19,4 @@ them, don't hesitate to ask. We're here to help! This is simply a reminder of wh
 
 - [] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
 - [] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
+- [] I have added artifacthub-pkg.yml file with the policy metadata and refer to the [template](https://github.com/artifacthub/hub/blob/master/docs/metadata/artifacthub-pkg.yml)
```
Comment (chipzoller) [inline .github/PULL_REQUEST_TEMPLATE.md:22]: "I have added the artifacthub-pkg.yml file and have verified it is complete and correct."

File: README.md [hunk 9]
```diff
@@ -58,6 +60,53 @@ spec:
             - Resource
 ```
 
+### Artifact Hub
+1. A dedicated folder must be created for each policy.
```
Comment (chipzoller) [inline README.md:64]: New line always between heading and start of content.

Comment (chipzoller) [reply README.md:64]: These three numbers should go above as bullets under the Contribution section.

File: README.md [hunk 10]
```diff
@@ -58,6 +60,53 @@ spec:
...
+2. The folder must be named the same as the policy.
+3. Use dashes for folder name and policy name instead of underscores.
+
+Add an `artifacthub-pkg.yml` metadata file to the folder. See an example metadata file for Kyverno policies below and customize per the comments.
+```yaml
```
Comment (chipzoller) [inline README.md:69]: Space between code blocks.

File: README.md [hunk 11]
```diff
@@ -58,6 +60,53 @@ spec:
...
+---
+name: backup-all-volumes # The name of the package (only alphanum, no spaces, dashes allowed) (required)
+version: 1.0.0 # Version of the policy (required)
+displayName: Backup All Volumes  # Display name of the policy (required)
+createdAt: "2023-03-29T00:00:00.000Z" # The date this package was created (RFC3339 layout) (required)
+description: >-
```
Comment (chipzoller) [inline README.md:75]: Please add comment here indicating the value should be taken from the relevant annotation.

File: README.md [hunk 12]
```diff
@@ -58,6 +60,53 @@ spec:
...
+      In order for Velero to backup volumes in a Pod using an opt-in approach, it
+      requires an annotation on the Pod called `backup.velero.io/backup-volumes` with the
+      value being a comma-separated list of the volumes mounted to that Pod. This policy
+      automatically annotates Pods (and Pod controllers) which refer to a PVC so that
+      all volumes are listed in the aforementioned annotation if a Namespace with the label
+      `velero-backup-pvc=true`.
+    # A short description of the package (required)
```
Comment (chipzoller) [inline README.md:82]: Remove

File: README.md [hunk 13]
```diff
@@ -58,6 +60,53 @@ spec:
...
+digest: 60ca548c88fc3e43db880bd5e466e4fa02af13b7f97c652aee59cb13ca9404e5 
+# The SHA256 hash String that uniquely identifies this package version (optional)
```
Comment (chipzoller) [inline README.md:84]: Add this as on the same line as above.

File: README.md [hunk 14]
```diff
@@ -58,6 +60,53 @@ spec:
...
+install: |-
+    ```shell
+    kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/velero/backup-all-volumes/backup-all-volumes.yaml
+     ```
+     # The installation instructions for the package (optional)
```
Comment (chipzoller) [inline README.md:89]: Put on same line as install.

File: README.md [hunk 15]
```diff
@@ -58,6 +60,53 @@ spec:
...
+keywords:
+  - velero
+  - kyverno
+  # Keywords associated with the package (optional)
```
Comment (chipzoller) [inline README.md:93]: Move to keywords line. Keywords should always have "kyverno" and whatever the value of the value of the `policies.kyverno.io/category` annotation.

File: README.md [hunk 16]
```diff
@@ -58,6 +60,53 @@ spec:
...
+readme: |
+  In order for Velero to backup volumes in a Pod using an opt-in approach, it
+  requires an annotation on the Pod called `backup.velero.io/backup-volumes` with the
+  value being a comma-separated list of the volumes mounted to that Pod. This policy
+  automatically annotates Pods (and Pod controllers) which refer to a PVC so that
+  all volumes are listed in the aforementioned annotation if a Namespace with the label
+  `velero-backup-pvc=true`.
+
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+# The README of the package (optional)
```
Comment (chipzoller) [inline README.md:103]: Same as previous comments.

Comment (chipzoller) [reply README.md:103]: readme should be same as `policies.kyverno.io/description` annotation plus the last sentence as a static value.

File: README.md [hunk 17]
```diff
@@ -58,6 +60,53 @@ spec:
...
+annotations:
+  kyverno/category: "Velero"
+  kyverno/kubernetesVersion: "1.25"
+  kyverno/subject: "Pod, Annotation"
+# Annotations associated with the package (optional)
```
Comment (chipzoller) [inline README.md:108]: Same as previous. Tell them where to get these annotations. Point them here: https://artifacthub.io/docs/topics/annotations/kyverno/

File: README.md [hunk 18]
```diff
@@ -58,6 +60,53 @@ spec:
...
+```
```
Comment (chipzoller) [inline README.md:109]: New line after code blocks.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Comment by chipzoller** [comment]
And you still have not addressed the second checkbox "All folders contain policies with the same name."

**Comment by guohaoyu110** [comment]
@chipzoller   please review, thank you.

File: README.md [hunk 19]
```diff
@@ -27,6 +27,15 @@ Anyone and everyone is welcome to write and contribute Kyverno policies! We have
...
 
+* Create the `artifacthub-pkg.yml` file in the same directory as your policy. See the Artifact Hub section below for more details on its contents.
+
+* A dedicated folder must be created for each policy.
+
+* The folder must be named the same as the policy.
+
+* Use dashes for folder name and policy name instead of underscores.
+
+
```
Comment (chipzoller) [inline README.md:38]: Remove excess empty lines.

File: README.md [hunk 20]
```diff
@@ -58,6 +67,41 @@ spec:
...
+displayName: Backup All Volumes  # Display name of the policy (required)
+createdAt: "2023-03-29T00:00:00.000Z" # The date this package was created (RFC3339 layout) (required)
+description: >-
+# The description value should be taken from the relevant annotation policies.kyverno.io/description
+digest: 60ca548c88fc3e43db880bd5e466e4fa02af13b7f97c652aee59cb13ca9404e5 # The SHA256 hash String that uniquely identifies this package version (optional)
+install: |- # The installation instructions for the package (optional)
+    ```shell
+    kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/velero/backup-all-volumes/backup-all-volumes.yaml
+     ```   
+keywords: # Keywords should always have "kyverno" and whatever the value of the value of the policies.kyverno.io/category annotation. (optional)
```
Comment (chipzoller) [inline README.md:87]: Duplicate "the value"

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Comment by guohaoyu110** [comment]
@chipzoller ok.

File: README.md [hunk 21]
```diff
@@ -58,6 +66,41 @@ spec:
...
+
+Add an `artifacthub-pkg.yml` metadata file to the folder. See an example metadata file for Kyverno policies below and customize per the comments.
+
+```yaml
+---
+name: backup-all-volumes # The name of the package (only alphanum, no spaces, dashes allowed)
+version: 1.0.0 # Version of the policy
+displayName: Backup All Volumes  # Display name of the policy
+createdAt: "2023-03-29T00:00:00.000Z" # The date this package was created (RFC3339 layout)
+description: >-
```
Comment (chipzoller) [inline README.md:79]: Please add the actual description here just like you've done below for the `readme` field.

File: README.md [hunk 22]
```diff
@@ -58,6 +66,41 @@ spec:
...
+readme: | # readme should be same as policies.kyverno.io/description annotation plus the last sentence as a static value.
+  In order for Velero to backup volumes in a Pod using an opt-in approach, it
+  requires an annotation on the Pod called `backup.velero.io/backup-volumes` with the
+  value being a comma-separated list of the volumes mounted to that Pod. This policy
+  automatically annotates Pods (and Pod controllers) which refer to a PVC so that
+  all volumes are listed in the aforementioned annotation if a Namespace with the label
+  `velero-backup-pvc=true`.
+
+  Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+annotations: # Annotations values should be the same as policies.kyverno.io annotations and get annotations from https://artifacthub.io/docs/topics/annotations/kyverno/
```
Comment (chipzoller) [inline README.md:98]: Change comment to:

"See the annotations guide on Artifact Hub here: https://artifacthub.io/docs/topics/annotations/kyverno/"

**Review by chipzoller — CHANGES_REQUESTED** [review]
In addition to these last few changes, I find that you still have not located and changed all policies to have names equal to their folders. One such example in /other/concatenate-configmaps

**Comment by guohaoyu110** [comment]
> In addition to these last few changes, I find that you still have not located and changed all policies to have names equal to their folders. One such example in /other/concatenate-configmaps

This is the script i write. I am pretty sure this time I have checked all the folders and yaml files. 

#!/bin/bash

# 遍历当前目录下的所有子文件夹
for dir in */; do
    # 获取文件夹名称
    folder_name=$(basename "$dir")

    # 构建预期的 .yaml 文件名
    yaml_file="${dir}${folder_name}.yaml"

    # 检查文件是否存在
    if [ -e "$yaml_file" ]; then
        echo "pass"
    else
        echo "In ${dir} cannot find the same name .yaml file"
    fi
done

**Comment by guohaoyu110** [comment]
@chipzoller  please check, thank you so much.

**Review by chipzoller — APPROVED** [review]
Looks good.

**Pull request merged**